286.py

```diff
@@ -0,0 +1,64 @@
+# class Solution:
+#     def wallsAndGates(self, rooms: List[List[int]]) -> None:
+#         """
+#         Do not return anything, modify rooms in-place instead.
+#         """
+#         self.rooms = rooms
+#         self.M, self.N = len(rooms), len(rooms[0])
+#         self.INF = 2147483647
+
+#         for i in range(self.M):
+#             for j in range(self.N):
+#                 if rooms[i][j] == self.INF:
+#                     rooms[i][j] = self.distanceToNearestGate(i, j)
+
+
+#     def distanceToNearestGate(self, start_row: int, start_col: int) -> int:
+#         queue = collections.deque([(start_row, start_col)])
+#         distance = [[0] * self.N for _ in range(self.M)]
+#         DIRS = ((1, 0), (-1, 0), (0, 1), (0, -1))
+
+#         while queue:
+#             row, col = queue.popleft()
+
+#             for dr, dc in DIRS:
+#                 r, c = row + dr, col + dc
+
+#                 if r < 0 or r >= self.M or c < 0 or c >= self.N or self.rooms[r][c] == -1 or distance[r][c] != 0:
+#                     continue
+
+#                 distance[r][c] = distance[row][col] + 1
+
+#                 if self.rooms[r][c] == 0:
+#                     return distance[r][c]
+
+#                 queue.append((r, c))
+
+#         return self.INF
+
+# Time Limit Exceeded.
+
+class Solution:
+    def wallsAndGates(self, rooms: List[List[int]]) -> None:
+        M, N = len(rooms), len(rooms[0])
+        EMPTY, GATE = 2147483647, 0
+        queue = collections.deque([])
+
+        for i in range(M):
+            for j in range(N):
+                if rooms[i][j] == GATE:
+                    queue.append((i, j))
+
+        DIRS = ((1, 0), (-1, 0), (0, 1), (0, -1))
+
+        while queue:
+            row, col = queue.popleft()
+
+            for dr, dc in DIRS:
+                r, c = row + dr, col + dc
+
+                if r < 0 or r >= M or c < 0 or c >= N or rooms[r][c] != EMPTY:
```

Comment:
It's probably better to invert the logic and use the pythonic way for range checks:

```python
if 0 <= r < R and 0 <= c < C and rooms[r][c] == EMPTY:
    # do something 
```

Reply:
I kind of disagree, usually bail early and save an indentation is the way to go.

Reply:
This is not a fail early situation. This block is essentially doing this:

```python
if not A:
    continue
else:
    # execute S1
    # execute S2
```
Isn't it equivalent of this?

```python
if A:
    # execute S1
    # execute S2
```
Indentation-wise, there is no difference, either. I'll leave it to you to decide which one is less fragmented and more readable.